text/3607-enum-discriminant.md

```diff
@@ -0,0 +1,413 @@
+let __self_tag = self.enum#discriminant;
+let __arg1_tag = other.enum#discriminant;
+::core::cmp::PartialOrd::partial_cmp(&__self_tag, &__arg1_tag)
+```
+is much easier.
+
+Outside of macros, something like
+```rust
+discriminant(&foo)
+```
+(which requires a `use std::mem::discriminant;`)
+isn't that different from
+```rust
+foo.enum#discriminant
+```
+
+And of course you can always make a function to give it a shorter name -- or write
+a proc macro to generate that function -- if you so wish.
+
+## Why just `pub(in self)`?
+
+The primary use case that led to this RFC is using it in `derive` macros, where
+`pub(in self)` is entirely sufficient.
+
+And by being only private, it avoids forcing any semver promises on library authors.
+
+Today, as a library author, you can reorder the variants in an enum should you so wish,
+or in a `#[non_exhausive]` enum add new ones in the middle.  There's no way for
+the users of your library to care about the order in which you defined the variants
+(unless you make other documented promises) -- especially if you never `derive(PartialOrd)`.
+
+Any library author who wishes to provide discriminant stability can always write
+a function to expose those discriminants, trivially implemented using this feature.
+
+## Why expose it via `.`?
+
+I like it behaving kinda like a field.  For example, having auto-deref like a field
+means you don't need to worry about whether you actually have a `&&Enum` in a `filter`
+or you actually have a `Box<Enum>` or whatever.
+
+Of course, if the `enum` is `repr(C)`, then the discriminant [is a field][RFC2195]
+in the guaranteed FFI layout, so thinking of it kinda like a field isn't too weird.
+
+There has also been talk of *compressed* or *move-only* fields where getting the
+address is disallowed so that Rust can run arbitrary logic whenever they're accessed
+and thus have the freedom to do more layout optimizations than are otherwise possible.
+Should we have something like that, then it's again not unreasonable to think of it
+as a field that sometimes has particularly fancy layout optimization.
+
+[RFC2195]: https://rust-lang.github.io/rfcs/2195-really-tagged-unions.html
+
+## What about if it was a magic method instead?
+
+It could be.  But it would still need to be something that doesn't cause name
+resolution failures for other methods that people might already have written.
+
+So I don't think that the extra `()` on it would really improve things.
+
+## Why not allow writing to the discriminant?
+
+The semantics for that get really complicated, especially for `enum`s in `repr(Rust)`
+that don't have a guaranteed layout, and even more so those that get layout-optimized.
+
+Maybe one day it could be allowed, but for now this RFC sticks only things that
+can be allowed in safe code without worries.
+
+## Couldn't this be a magic macro?
+
+Sure, it could, like `offset_of!`.
+
+I don't think `enum_discriminant!(foo)` is really better than `foo.enum#discriminant`, though.
```

Comment:
While I do see the argument, I do think there is a point in favour of `discriminant_of!` - the spending of weirdness budget.
Magic macros are nothing new in rust at this point (`addr_of!`, `offset_of!`), and in theory, both of those could have been implemented with magic fields (or even bespoke syntax, eg. `&raw`).
I think the RFC needs a strong argument on why this weirdness budget should be spent on this feature. Implementing custom derives is presumably a lot more rare than working with `addr_of!` in pointer handling code.

Reply:
another benefit of magic macros: you can easily get the discriminant of a variant without having to have/create a value:
```rust
enum E {
    A(Arc<Mutex<String>>) = 3,
    B,
}

const _: () = assert!(discriminant_of!(E::A) == 3);
```

Reply:
@programmerjake such magic macro cannot exist in this form, it will not able to disambiguate the follow case

```rust
#[repr(u8)]
enum F {
    A { a: u8 } = 1,
    B = 2,
}

impl F {
    const A: Self = Self::B;
}

const _: () = assert!(discriminant_of!(F::A) == 2);
```

since the primary motivation of `.enum#discriminant` is to support `#[derive(PartialOrd)]`, the argument to that macro is usually an expression, meaning `F::A` will be referring to that associated const 

Yes you could make a separate macro to only handle `discriminant_of_variant!(F::A) == 1`, but it will not be able to solve the problems this RFC is handling.

Reply:
Personally I would prefer custom syntax for new fundamental concepts, rather than magic macros. Any kind of macros makes the code harder to analyze, both automatically and for humans. One needs to ensure that the magic macro wasn't shadowed by a non-magic user macro (which can be quite nontrivial, given the textual scoping possible with `macro_rules!`), and to remember that some macros actually act as built-in syntax. Conversely, special syntax cannot be overridden, and anyone who can parse it likely can remember that it has special semantics.

On the other hand, magic syntax doesn't have a natural place to attach documentation, so it may be less discoverable and harder to learn.